tools/doc-gen/index.ts

```diff
@@ -0,0 +1,377 @@
+ * @param ruleInfo
+ * @returns
+ */
+async function readRuleData(ruleInfo: RuleFileInfo): Promise<RuleData> {
+  // Parse rule test file
+  const specRawSource = readFileSync(ruleInfo.spec, 'utf-8');
+  const specSource = ts.createSourceFile(
+    ruleInfo.spec,
+    specRawSource,
+    ts.ScriptTarget.Latest
+  );
+
+  const validCases: ValidCase[] = [];
+  const invalidCases: InvalidCase[] = [];
+
+  const findTestCases = (node: ts.Node) => {
+    if (
+      ts.isPropertyAssignment(node) &&
+      ts.isIdentifier(node.name) &&
+      node.name.escapedText === 'valid' &&
+      ts.isArrayLiteralExpression(node.initializer)
+    ) {
+      // Found valid test cases array
+      const validCaseNodes = node.initializer.elements;
+      validCaseNodes.forEach((caseNode) => {
+        if (!ts.isObjectLiteralExpression(caseNode)) {
+          // not expected format
+          return;
+        }
+        const nameNode = caseNode.properties.find(
+          (prop) =>
+            ts.isPropertyAssignment(prop) &&
+            ts.isIdentifier(prop.name) &&
+            prop.name.escapedText === 'name'
+        );
+        const codeNode = caseNode.properties.find(
+          (prop) =>
+            ts.isPropertyAssignment(prop) &&
+            ts.isIdentifier(prop.name) &&
+            prop.name.escapedText === 'code'
+        );
+
+        const caseInfo: any = {};
+        if (
+          nameNode &&
+          ts.isPropertyAssignment(nameNode) &&
+          ts.isStringLiteral(nameNode.initializer)
+        ) {
+          caseInfo['name'] = nameNode.initializer.text;
+        } else {
+          // No case name provided, so abort
+          return;
+        }
+
+        if (
+          codeNode &&
+          ts.isPropertyAssignment(codeNode) &&
+          ts.isNoSubstitutionTemplateLiteral(codeNode.initializer)
+        ) {
+          caseInfo['code'] = codeNode.initializer.text;
+        }
+
+        if (caseInfo['code']) {
+          validCases.push(caseInfo);
+        }
+      });
+    }
+    if (
+      ts.isPropertyAssignment(node) &&
+      ts.isIdentifier(node.name) &&
+      node.name.escapedText === 'invalid' &&
+      ts.isArrayLiteralExpression(node.initializer)
+    ) {
+      // Found invalid test cases array
+      const invalidCaseNodes = node.initializer.elements;
+      invalidCaseNodes.forEach((caseNode) => {
+        if (ts.isObjectLiteralExpression(caseNode)) {
+          // Object format case
+          const nameNode = caseNode.properties.find(
+            (prop) =>
+              ts.isPropertyAssignment(prop) &&
+              ts.isIdentifier(prop.name) &&
+              prop.name.escapedText === 'name'
+          );
+          const codeNode = caseNode.properties.find(
+            (prop) =>
+              ts.isPropertyAssignment(prop) &&
+              ts.isIdentifier(prop.name) &&
+              prop.name.escapedText === 'code'
+          );
+
+          const caseInfo: any = {};
```

Comment:
evil any~, can we add a TODO: at the top to polish the types a bit

Reply:
I've cleaned up the `any`s 👍